### ISSUE TITLE
Rollback to snapshot 

### ISSUE THREAD

**Issue by canada4663** [manahl/arctic#153]
Folks -

In mass loading historical data / I have gotten some failures - segmentation faults- that I haven't been able to catch with try/except blocks.  

I wrote a custom "rollback" function as I couldn't see anything inherent in the library that allowed a rollback of a transaction.  Essentially I am taking a snapshot at each known "good" append - and if I have a fail on next append, I rollback and re-append.

I'm running into some kinks in the implementation - suspect I may not have fully understood how the pandas / version store is working.... So before I debug again, thought I would check to make sure I'm not missing an easy rollback feature.  

Thoughts?

Ryan

**Comment by bmoscon** [comment]
VersionStore support snap shots and versions. Either one lets you "go back in time" and view the data as it was. I guess I'm not sure why you felt the need to write your own -- what does VersionStore lack?

**Comment by bmoscon** [comment]
@canada4663 any update on this?

**Comment by canada4663** [comment]
sorry for missing this!  I'm pulling from my memory banks a bit, but the reason I couldn't just use the 'as_of' feature you have with the 'read' function as that this wasn't for reads... it was for appends.

What was happening I was I was appending ~5k symbols of data for decent size data frames (overall database is ~250GB).  I was getting segfaults that were making it difficult to figure out where to restart the batch - so I needed to roll back to a known good point in the data and start at that point.

If append had a 'as_of' feature, this would have been perfect - but I just hacked something together using metadata I was already putting in the data to determine last good point and deleted versions/chunks after that point.  Tried a few other things, but this worked well for me.

Crude and I know won't be guaranteed to be backwards compatible since I had to use some internals to get it done.  I thought about doing a PR for it, but since I used metadata to make the decision, it won't work for the general code base.

BTW - the segfaults were not an arctic or mongo problem, just the robustness of my simple loader script wasn't great.  the process was getting way too big in terms of memory consumption.  This was happening because however I was managing garbage collection of the data frames being read from CSV and written to arctic wasn't smart enough - when I read smaller ranges of data per process, terminate the process, and start up another loader instance, the segfaults go away.

**Comment by bmoscon** [comment]
Oh I see, you want to append to a previous version, not necessarily the last version, but any previous version?

**Comment by canada4663** [comment]
Exactly!

**Comment by bmoscon** [comment]
I've updated the title so its clearer :)

**Comment by jamesblackburn** [comment]
This sounds like it might be #228 . Does this look familiar?

If not, without the exception / error you were experiencing it's unclear what problem you were facing.

**Comment by bmoscon** [comment]
@jamesblackburn this isnt a bug, hes asking for a feature - to be able to append to an older version, and have the result be the new version. Its almost like you start with data "a", you append "b" then decide, no, thats wrong, take me back to "a" and append "c"

**Comment by jamesblackburn** [comment]
I'm not so sure...  the discussion above talks a lot about segfaults in his app, and rolling back versions. Moreover figuring out which version to roll back to (in an app) wouldn't be a trivial thing - so I'm not sure I see the utility of doing so. 

To me it sounds like the issue is that append may not be crash-safe / transactionally safe - which looks to be true.

**Comment by richardbounds** [comment]
I tried doing something like this when I wrote it originally, but I think I ended up running into a problem with one of the unique constraints - if only I could remember exactly what it was....

It does all get a bit complicated when you try to 'fork' the timeseries like this - it might be simpler just to do a read(version)/concat/write in the client app. If you're trying to avoid ever reading the whole timeseries into memory, you're probably in trouble already.

**Comment by bmoscon** [comment]
I agree with @richardbounds, there are easy ways with the API to work around this
